### PR TITLE
Add planner code for replicated clickhouse processes

### PR DESCRIPTION
**Pull request by andrewjstone**
Primarily, this commit includes a planning mechanism to allocate clickhouse keepers and servers to their in-service zones. This `ClickhouseAllocator` is intended to be created by the `BlueprintBuilder` after zones are added and expunged. It very carefully manages adding or removing only one keeper node at a time and there are a bunch of tests regarding this  behavior.

In order to know when to continue allocating new keepers or servers or expunging them, the allocator takes in the set of zones from the current builder run, the parent configuration from the last blueprint, and any inventory updates related to clickhouse keepers. We only pass in the `latest_clickhouse_keeper_membership` from the inventory. This is all that the allocator needs and also simplifies testing by precluding having to build a complete `Collection` just to test. While we pass in all `blueprint_zones` from the `BlueprintBuilder` to the `ClickhouseAllocator::new` constructor, to make production code simpler and safer to write, we mnaually manipulate the allocator to keep the tests simple, and not have to worry about generating `BlueprintZonesConfig` that we don't actually use.

None of this code is wired up to run, but all the planning logic should be capable of being hooked in easily. We'll also need to add DB models and queries, and write the executor, and inventory collection in follow ups.

### CODE REVIEW

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:154]:
```suggestion
        // Save the log index for next time
```

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:159]:
```suggestion
            // We're still trying to reach our desired state. We want to ensure
```

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (karencfv) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:68]: When would ClickhouseKeeperClusterMembership be `None`? (This question is just me trying to understand how reconfigurator works.)

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [inline nexus/types/src/deployment/clickhouse.rs:23]: Nit

```suggestion
    /// Clickhouse Server IDs must be unique and are handed out monotonically.
```

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [inline nexus/types/src/deployment/clickhouse.rs:97]: What do you think about using this constant instead of taking cluster_name as a parameter? 

 https://github.com/oxidecomputer/omicron/blob/049663724ecf779e43dad7908a16f07055a68a4e/clickhouse-admin/types/src/lib.rs#L18

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [inline nexus/types/src/deployment/clickhouse.rs:33]: At the moment we're hardcoding this value into the ClickHouse config templates. What would this field be used for?

https://github.com/oxidecomputer/omicron/blob/fef8616bef83cc21d16dc7951b3f48fa88a0b106/clickhouse-admin/types/src/config.rs#L209-L210

Should we wait to handle this value securely before adding it to this config struct? https://github.com/oxidecomputer/omicron/issues/3823

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [inline nexus/types/src/deployment/clickhouse.rs:77]: `is has it as part of its desired state`? 😄

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [inline nexus/types/src/deployment/clickhouse.rs:81]: This process makes sense to me!

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (karencfv) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:185]:
```suggestion
            // If we are not adding a keeper then we are done.
```

**Review by karencfv — COMMENTED** [review]
So great to see this coming alive! I can see how getting the keeper add/remove process right was incredibly tricky. Just a few questions and some typos, but looking good to me.

Would still wait for a review from someone that has a better grasp of reconfigurator than me though.

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs:97]: I don't think that's necessarily going to work in the future multi-rack world, but it might!  I can envision customers having separate clusters for different racks or fleets and wanting to distinguish between them possibly. Maybe we wouldn't tie the cluster name that way, but I wouldn't guarantee keeping them the same indefinitely.

In any case, I don't think it hurts to keep it flexible for now. For this configuration we are only going to set the name once and carry it forward indefinitely. At the call site we can pass in the constant.

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs:33]: I think the name 'secret' is a misnomer. It doesn't provide any real security guarantees. For those we'll end up using certificates on the control plane for authentication. The main usage is just to make sure that nodes run accidentally with access to each other's IPs aren't actually talked to. Quoting the [clickhouse replication docs](https://clickhouse.com/docs/en/architecture/replication):
> A secret is created for the cluster named cluster_1S_2R with the value mysecretphrase. The secret is shared across all of the remote servers in the environment to ensure that the correct servers are joined together.

So I think we'll end up just generating a UUID for this and stringifying it.

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:68]: It would be `None` if the latest inventory collection failed to reach all keepers. This would most likely happen due to a transient error like a network partition or reboot.

File: nexus/db-queries/src/db/datastore/inventory.rs
Comment (andrewjstone) [reply nexus/db-queries/src/db/datastore/inventory.rs]: Good idea! Done in [340046e](https://github.com/oxidecomputer/omicron/pull/6566/commits/340046eac678f712d813de8a6bcb7ed8fbb5435a)

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: Fixed in [340046e](https://github.com/oxidecomputer/omicron/pull/6566/commits/340046eac678f712d813de8a6bcb7ed8fbb5435a)

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs]: Good catch. Fixed in [340046e](https://github.com/oxidecomputer/omicron/pull/6566/commits/340046eac678f712d813de8a6bcb7ed8fbb5435a)

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: Fixed.

**Comment by andrewjstone** [comment]
> So great to see this coming alive! I can see how getting the keeper add/remove process right was incredibly tricky. Just a few questions and some typos, but looking good to me.
> 
> Would still wait for a review from someone that has a better grasp of reconfigurator than me though.

Thanks for the quick review @karencfv. Much appreciated!

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (karencfv) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:70]: And since reconfigurator is constantly polling (is this right?), it's not a big deal if this information is empty every so often?

File: nexus/types/src/deployment/clickhouse.rs
Comment (karencfv) [reply nexus/types/src/deployment/clickhouse.rs:33]: Ahhh, yeah, ok that makes sense.

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:520]: I think this is what I was wondering about above re: `leader_committed_log_index`. It looks like in this case we've only changed the blueprint, but we still need the committed log index to be going up to make the right planning decisions?

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (jgallagher) [inline nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:848]: These tests are awesome <3

File: nexus/types/src/deployment.rs
Comment (andrewjstone) [reply nexus/types/src/deployment.rs:350]: Thank you for the careful review! I think a strong type makes the most sense here, although I'm hoping it won't ripple much throughout the codebase.

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs:68]: Good eyes. Yes, you are correct. I will be more explicit.

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:203]: Good eye again! I think this was not updated when I ended up changing the name of that variable.

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:40]: I basically agree with you, but I find the name `ShouldBeRunning` tedious to use in variables. I should probably change this for consistency though.

Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:40]: Also agreed. Maybe we need a better name there (but still leave "in service" with its RFD-defined meaning)?

Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:40]: Yeah, I don't want to change the RFD definitions; for now at least.

Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:40]: What about `ActiveClickhouseZones` ?

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs]: Fixed in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/inventory.rs
Comment (andrewjstone) [reply nexus/types/src/inventory.rs:170]: Thanks! Changed in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/inventory.rs
Comment (andrewjstone) [reply nexus/types/src/inventory.rs]: Fixed in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: Changed in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: Cleaned up in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: Fixed in [f9c3307](https://github.com/oxidecomputer/omicron/pull/6566/commits/f9c3307457d4fabd6b385e41ba35b3b0df57e71e)

File: nexus/types/src/deployment.rs
Comment (andrewjstone) [reply nexus/types/src/deployment.rs:350]: Alright, I changed my mind against a strong type. It will have too many ripple effects and I'm not sure it actually makes things clearer. I started to lean towards scoping this function to the Blueprint, but a method already exists with the same name. I still like this solution better than the free function though, so I'm going to rename it to `all_bp_zones`. If you think of a better name, I am happy to hear it.

Comment (andrewjstone) [reply nexus/types/src/deployment.rs]: Changed in [8e2ee68](https://github.com/oxidecomputer/omicron/pull/6566/commits/8e2ee686172a7e68352fd2c72cb777179a1f5af5)

Comment (andrewjstone) [reply nexus/types/src/deployment.rs]: Damnit, forgot to cleanup the comment. Doing that now.

Comment (andrewjstone) [reply nexus/types/src/deployment.rs]: Comment fixed in [23185e8](https://github.com/oxidecomputer/omicron/pull/6566/commits/23185e831bb4593748ec19550bf7bf10cdd6f98d)

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs]: I changed this behavior in [9401e22](https://github.com/oxidecomputer/omicron/pull/6566/commits/9401e22688052011cb70500791fd0f50fbc60edb) and changed that point in the test to ensure it works as expected. In the other cases I looked at it seems unnecessary to change the tests.

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs]: You were right as discussed in https://github.com/oxidecomputer/omicron/pull/6566#discussion_r1761334847

I changed the behavior in [9401e22](https://github.com/oxidecomputer/omicron/pull/6566/commits/9401e22688052011cb70500791fd0f50fbc60edb)

File: nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs
Comment (andrewjstone) [reply nexus/reconfigurator/planning/src/blueprint_builder/clickhouse.rs:40]: I ended up splitting the difference: I renamed the type to `ClickhouseZonesThatShouldBeRunning`, but named
the member variable `active_clickhouse_zones` for readability in [b5fc8bb](https://github.com/oxidecomputer/omicron/pull/6566/commits/b5fc8bbe20fbf94bba9b671992faf33ce993ba2b)

**Comment by andrewjstone** [comment]
Thanks for the thorough review @jgallagher. I believe I resolved everything you pointed out.

File: nexus/types/src/deployment.rs
Comment (jgallagher) [reply nexus/types/src/deployment.rs]: > I still like this solution better than the free function though, so I'm going to rename it to all_bp_zones. If you think of a better name, I am happy to hear it.

Maybe since the scoped function doesn't actually accept a blueprint, it could be `zones_filtered` or something like that?

File: nexus/types/src/deployment/clickhouse.rs
Comment (jgallagher) [inline nexus/types/src/deployment/clickhouse.rs:26]:
```suggestion
    /// Clickhouse Keeper IDs must be unique and are handed out monotonically.
```

**Review by jgallagher — APPROVED** [review]
Ship it!

File: nexus/types/src/deployment.rs
Comment (andrewjstone) [reply nexus/types/src/deployment.rs]: Ah, I like that. I changed it to `filtered_zones` though.

File: nexus/types/src/deployment/clickhouse.rs
Comment (andrewjstone) [reply nexus/types/src/deployment/clickhouse.rs]: I have an amazing capacity for typos.

**Pull request merged**
